### PR TITLE
beam-refactor: add start/stop fields to DimIndex

### LINKED ISSUE #367: Augment Index index with start & stop
As per discussion today, we want to do the following in `patterns.py`

```diff
@dataclass(frozen=True, order=True)
class DimIndex:
    name: str
    index: int
    sequence_len: int
    operation: CombineOp
+   start: Optional[int]
+   stop: Optional[int]
```
`start` and `stop` represent the explicit absolute array index location of the data in this chunk in the full dataset. So to insert the data, you would do (assuming 1D)
```python
target[dim_index.start:dim_index.stop] = data_from_chunk
```

We also want to write a standalone function of the form

```python
augment_index_with_start_stop(index: DimIndex, sequence_lens: List[int]) -> DimIndex:
    # take an index _without_ start / stop and add them based on the lens defined in sequence_lens
    assert len(sequence_lens) == index.sequence_len
    start = sum(sequence_lens[:index.index])
    stop = start + sequence_lens[index.index]
    # return a new dimindex because it's frozen
``` 
`sequence_lens` is the list of the lens of _all the items in the sequence_ including this one.

This can be unit tested with synthetic data.

### PR DESCRIPTION
**Pull request by cisaacstern**
Closes #367 when complete

### CODE REVIEW

**Comment by cisaacstern** [comment]
@rabernat, I believe this covers the work scoped in #367. If I've overlooked something, let me know and I can amend.

File: pangeo_forge_recipes/patterns.py [hunk 1]
```diff
@@ -101,6 +110,23 @@ class Index(FrozenSet[DimIndex]):
 FilePatternIndex = Index
 
 
+def augment_index_with_start_stop(index: DimIndex, sequence_lens: List[int]) -> DimIndex:
```
Comment (rabernat) [inline pangeo_forge_recipes/patterns.py:113]:
```suggestion
def augment_index_with_start_stop(index: DimIndex, item_lens: List[int]) -> DimIndex:
```

File: pangeo_forge_recipes/patterns.py [hunk 2]
```diff
@@ -101,6 +110,23 @@ class Index(FrozenSet[DimIndex]):
...
+    """Take an index _without_ start / stop and add them based on the lens defined in sequence_lens.
+
+    :param index: The ``DimIndex`` instance to augment.
+    :param sequence_lens: A list of integer sequence lengths for all inputs.
```
Comment (rabernat) [inline pangeo_forge_recipes/patterns.py:117]:
```suggestion
    :param item_lens: A list of integer lengths for all items in the sequence.
```

File: pangeo_forge_recipes/patterns.py [hunk 3]
```diff
@@ -101,6 +110,23 @@ class Index(FrozenSet[DimIndex]):
...
+    """
+
+    assert len(sequence_lens) == index.sequence_len
```
Comment (rabernat) [inline pangeo_forge_recipes/patterns.py:120]:
```suggestion
    assert len(item_lens) == index.sequence_len
```

File: pangeo_forge_recipes/patterns.py [hunk 4]
```diff
@@ -101,6 +110,23 @@ class Index(FrozenSet[DimIndex]):
...
+    start = sum(sequence_lens[: index.index])
```
Comment (rabernat) [inline pangeo_forge_recipes/patterns.py:121]:
```suggestion
    start = sum(item_lens[: index.index])
```

File: pangeo_forge_recipes/patterns.py [hunk 5]
```diff
@@ -101,6 +110,23 @@ class Index(FrozenSet[DimIndex]):
...
+    stop = start + sequence_lens[index.index]
```
Comment (rabernat) [inline pangeo_forge_recipes/patterns.py:122]:
```suggestion
    stop = start + item_lens[index.index]
```

**Review by rabernat — COMMENTED** [review]
This looks awesome! In reviewing I realized we should probable rename a variable.

**Comment by rabernat** [comment]
Isn't there an "add suggestion to batch" option?

**Comment by cisaacstern** [comment]
Oh good to know for next time! I hadn't been aware of that option. 🙏

**Pull request merged**